aya/src/maps/ringbuf.rs

```diff
@@ -0,0 +1,355 @@
+    pub(crate) fn new(map: T) -> Result<Self, RingBufferError> {
+        // Check that the map is a ringbuf
+        let map_type = map.obj.def.map_type;
+        if map_type != BPF_MAP_TYPE_RINGBUF as u32 {
+            return Err(MapError::InvalidMapType { map_type }.into());
+        }
+
+        // Determine page_size, map_fd, and set mask to map size - 1
+        let page_size = unsafe { sysconf(_SC_PAGESIZE) } as usize;
+        let map_fd = map.fd_or_err().map_err(RingBufferError::from)?;
+        let mask = (map.obj.def.max_entries - 1) as usize;
+
+        // Map writable consumer page
+        let consumer_page = unsafe {
+            mmap(
+                ptr::null_mut(),
+                page_size,
+                PROT_READ | PROT_WRITE,
+                MAP_SHARED,
+                map_fd,
+                0,
+            )
+        };
+        if consumer_page == MAP_FAILED {
+            return Err(RingBufferError::ConsumerMMapError {
+                io_error: io::Error::last_os_error(),
+            });
+        }
+
+        // From kernel/bpf/ringbuf.c:
+        // Each data page is mapped twice to allow "virtual"
+        // continuous read of samples wrapping around the end of ring
+        // buffer area:
+        // ------------------------------------------------------
+        // | meta pages |  real data pages  |  same data pages  |
+        // ------------------------------------------------------
+        // |            | 1 2 3 4 5 6 7 8 9 | 1 2 3 4 5 6 7 8 9 |
+        // ------------------------------------------------------
+        // |            | TA             DA | TA             DA |
+        // ------------------------------------------------------
+        //                               ^^^^^^^
+        //                                  |
+        // Here, no need to worry about special handling of wrapped-around
+        // data due to double-mapped data pages. This works both in kernel and
+        // when mmap()'ed in user-space, simplifying both kernel and
+        // user-space implementations significantly.
+        let producer_pages = unsafe {
+            mmap(
+                ptr::null_mut(),
+                page_size + 2 * (mask + 1),
+                PROT_READ,
+                MAP_SHARED,
+                map_fd,
+                page_size as i64,
+            )
+        };
+        if producer_pages == MAP_FAILED {
+            return Err(RingBufferError::ProducerMMapError {
+                io_error: io::Error::last_os_error(),
+            });
+        }
+
+        Ok(RingBuf {
+            _map: Arc::new(map),
+            map_fd,
+            data_ptr: unsafe { (producer_pages as *mut u8).add(page_size) },
+            consumer_pos_ptr: consumer_page as *mut _,
+            producer_pos_ptr: producer_pages as *mut _,
+            page_size,
+            mask,
+        })
+    }
+
+    /// Retrieve an event from the ring, pass it to the callback, mark it as consumed, then repeat.
```

Comment:
Meta comment about docs: for consistency the format we use is the same as rust's libstd. The format is roughly:

```
First line describes briefly and in 3rd person what the function does.

Second paragraph expands on details, mentioning in discursive form the arguments and the return value.

More paragraphs might be needed to explain caveats and edge cases.

# Example 
...
```

Sentences always end with periods. Paragraphs are separated by an empty line.

Reply:
I added some newlines. I don't write explanation for most of the arguments because most of the time they are the only argument and the meaning is obvious, but if you have any feedback around how docs is written in this PR, let me know.